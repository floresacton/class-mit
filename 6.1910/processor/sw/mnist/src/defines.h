#ifndef DEFINES_H_
#define DEFINES_H_

// Typedefs to make the bit size easier to interpret
typedef signed char int8_t;
typedef unsigned char uint8_t;
typedef signed short int16_t;
typedef unsigned short uint16_t;
typedef signed int int32_t;
typedef unsigned int uint32_t;
typedef long long int64_t;
typedef unsigned long long uint64_t;

#endif