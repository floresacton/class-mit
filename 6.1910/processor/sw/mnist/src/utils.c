#include "utils.h"

void print_string(char* s) {
    while (*s != '\0') {
        print_char(*s);
        s++;
    }
}

const int INT_PRINT_LIMIT = 10;
void print_int(int a) {
    int bases[INT_PRINT_LIMIT];
    bases[0] = 1;
    int i = 0;
    for (i = 1; i < INT_PRINT_LIMIT; i++) {
        bases[i] = bases[i - 1] * 10;
    }
    i = 0;
    if (a == 0) {
        print_char('0');
    } else if (a == 0x80000000) {
        print_string("-2147483648");
    } else {
        if (a < 0) {
            print_char('-');
            a = -a;
        }
        int j = INT_PRINT_LIMIT - 1;
        while (a < bases[j]) j--;
        while (j >= 0) {
            int d = 0;
            while (a >= bases[j]) {
                a -= bases[j];
                d ++;
            }
            print_char('0' + d);
            j--;
        }
    }
}

void print_int_hex(unsigned int a) {
    const char hex_lookup[] = "0123456789abcdef";
    int len = 0;

    if (a == 0) len = 1;
    else for (unsigned int n = a; n; n >>= 4) len++;

    for (int i = len - 1; i >= 0; i--)
        print_char(hex_lookup[(a >> (4 * i)) & 0xf]);
}