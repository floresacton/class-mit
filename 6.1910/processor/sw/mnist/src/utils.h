#ifndef UTILS_H_
#define UTILS_H_


// this is linked  from the crt0.S
void print_char(char x);

void print_string(char* s);

void print_int(int a);

void print_int_hex(unsigned int a);

#endif