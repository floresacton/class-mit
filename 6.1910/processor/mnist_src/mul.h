#ifndef MUL_H_
#define MUL_H_

#include "defines.h"

// GCC uses this unsigned multiplication function to override the * operator for unsigned/signed int multiply
uint32_t __mulsi3 (uint32_t a, uint32_t b) {
    uint32_t result = 0;

    // Examples of how to inline new instructions into C
    // read more at : https://sourceware.org/binutils/docs-2.34/as/RISC_002dV_002dFormats.html
    // and  https://gcc.gnu.org/onlinedocs/gcc/Extended-Asm.html
    //
    // asm volatile (".insn r 0b0110011, 0b000, 0b0000001, %0, %1, %2" : "=r" (result) : "r"(a), "r"(b)); 
 
    while (b) {
        if (b & 1) {
            result += a;
        }
        a = a << 1;
        b = b >> 1;
    }
    return result;

}

// Same as above but for 64 bit multiplication

// WARNING: Using unsigned 32x32 bit multiplication here will cause signed mulitplicaitons
// to fail! Since only 64x64 bit multiplicaiton will work, we suggest you implement and
// call a separate helper function instead which computes signed 32x32 bit multiplicaiton directly.

// If you would like to test your function with the provided testbench,
// make sure to change call to * in sw/mnist/mul64_test.c
uint64_t __muldi3 (uint64_t a, uint64_t b) {
    uint64_t result = 0;
    
    while (b) {
        if (b & 1) {
            result += a;
        }
        a = a << 1;
        b = b >> 1;
    }
    return result;

}

// SUGGESTED: Integrate into model.c and then implement in hardware
int32_t packmul(uint32_t input_i, uint32_t filter_i) {
    // input_i and filter_i each hold four adjacent int8_t numbers cast to a single uint32_t

    // NOTE: Packed elements of input buffer should be treated as unsigned.
    uint8_t a[4]; *(uint32_t*)a = input_i;
    int8_t b[4]; *(uint32_t*)b = filter_i; 

    // NOTE: software implementation of packmul relies on working 32 bit multiplication
    return (
        ((int32_t)a[0] * (int32_t)b[0]) +
        ((int32_t)a[1] * (int32_t)b[1]) +
        ((int32_t)a[2] * (int32_t)b[2]) +
        ((int32_t)a[3] * (int32_t)b[3])
    );
}

#endif