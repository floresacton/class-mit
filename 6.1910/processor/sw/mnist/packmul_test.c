#include "src/defines.h"
#include "src/utils.h"
#include "src/mul.h"

const volatile uint32_t a[] = {1351735424,1452906908,1995725930,1923404158,4020762764,1590695523,3796268576,313980746,927455040,1879963342,3342556673,1749807198,3304154373,2551634782,4065934357,1832567514,282093348,1565880994,1999420519,1267338533,1497917785,1953134459,4195716201,2505781775,3642327475,2636666266,3094522662,1761762889,3825931988,1055432037,203104764,2244679349,1364308297,3626522775,3770389804,2226316663,483011732,3836128094,3216343863,32477355,3370121715,2643108563,18767643,3847011960,3770871272,1172687165,4167617221,493451948,732893559,1572741726,1736234271,3571884529,3524421636,1944778137,1556815405,1925585144,2187221443,151264043,2900690994,1181093746,1192575604,2146758244,2079677389,2607960961,566795218,2874435025,1707958342,500023378,1344051070,2147215850,1948951637,464899997,4197912353,2226123776,3429358605,737796484,3364929790,743520928,1477664966,725439180,2860829339,1608613759,2660138818,4160828677,1866908966,3177519147,518422069,1495883903,2488485988,2052708446,1956547278,796460037,749428652,2261970775,2019112651,1225964212,852027437,1995465699,903918819,2217520692,2156699622,1942487345,2176626586,1106146040,1607134236,3219894972,1672327211,820533319,3329885456,3286406661,2480828836,4162286134,229534031,3678831644,80004865,1663082212,3153457107,276590231,1779349259,828021828,672462341,2485607260,3565894284,1029801201,2109228175,3517932156,768696856,2367148900,3194733176,4079325223,1268168767,3889875331,4003203200,2250815910,604760474,2408669363,1867240401,2839330654,2473089493,3399025590,1957252322,2582396901,1958537789,1564553869,1525441353,3626417493,3531335024,3367652374,3857206555,2120166227,1287909581,209952823,2187753277,983433585,2887898175,2937190132,3416347938,1276945709,1033009048,1672225559,3707271280,4193932264,2562418443,1099909766,2624314607,3382189989,686535488,3203147639,1858127193,945511383,3319036558,158167437,1975563657,2639313085,350712935,396422523,3749426739,3999898055,2166578146,4031398239,1955383838,2315379988,1498100227,286835136,4017890045,4089730256,1673518703,1776379219,3000247968,238403481,1416070045,4025823525,3452374126,203951749,1549180473,1802324022,85707792,2192129997,2759065332,639138082};
const volatile uint32_t b[] = {2449732032,514597339,2038455446,1203561948,1198667812,4148776890,2227890385,2439613732,909339715,1046709913,1623673218,1869257233,2857468264,3443077045,145030982,2238214436,3053920985,3810579489,2945589868,1385074179,1733168506,3659571766,3093570828,4234122981,4178339245,2025820983,1263628989,619516577,2327155498,3135822364,1388044452,4007309810,1588044244,1385425608,561837058,128636962,1182803727,1950117664,3671065766,1786470108,509255833,1231316961,299903318,4030421978,3935574232,3621629166,2520305975,3575785006,1684471745,1959349349,1725470179,1112350300,2372196701,2431691672,345083032,1391797504,3931422873,1232408018,3254045544,1571484118,74971289,3114089710,3850841890,1878731255,76909930,1950482513,3537547168,2574320903,2140694887,2890665070,3921520993,3151720246,2132691636,2415410428,2649817863,359495035,3589479830,1839675886,1631862215,2728409700,402544254,1856270065,3592235670,125810070,2334612739,3620421491,1690481352,2689612859,106355124,1249537117,3997349844,794212634,2193272345,2289929615,1129473472,294918828,3393737134,4062556219,2488906166,4290536453,3966841746,3943979089,3533245167,4276763751,4261373733,1111368634,1698492678,4189079206,635856997,4051908387,2458161430,1017723788,726555332,3863189040,519386061,4237704394,634911918,953250452,422401842,3485807084,4206233415,1016502499,4155213073,3835587469,3324675435,1486890148,4249515205,391167683,3621119659,1736192460,1950639645,165633897,1264897266,1440568754,1827298563,1630632026,4183461300,3584148389,1560255364,1818309168,1306709618,3621413927,1321601416,1199988245,3959634205,347862435,179635108,4019706934,1374955928,1060456261,2176119675,3211233133,229560472,796338685,2155954106,3316252489,2060473740,4033345597,1522912483,2223157445,90545903,585730300,420178156,3727977853,2247535625,3877702469,2990463436,3653534073,1851569016,2382277044,3618057516,4157313446,2827185043,2043689522,3939837877,3360043080,363643022,1791323420,2536403120,2756115248,3653009758,1395793571,1414799707,1327061486,1087006891,3415197033,2469173745,1251842231,4109381452,3549452298,3919133109,715325166,1280339638,989395231,96008691,4178216533,3686243859,3869278766,953355575,2860860129};
    
int main(int argc, char* argv[]) {
    for(int test_no = 0; test_no < sizeof(a)/sizeof(a[0]); test_no ++)  {
        print_string("Input ");
        print_int(test_no);
        print_string(" - ");
        print_int(a[test_no]);
        print_string(" * ");
        print_int(b[test_no]);
        print_string(" = ");
        print_int(packmul(a[test_no],b[test_no]));
        print_string("\n");

    }
    return 0;
}
