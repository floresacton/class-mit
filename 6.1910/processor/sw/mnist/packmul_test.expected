Input 0 - 1351735424 * -1845235264 = -19797
Input 1 - 1452906908 * 514597339 = -10995
Input 2 - 1995725930 * 2038455446 = -19726
Input 3 - 1923404158 * 1203561948 = -13237
Input 4 - -274204532 * 1198667812 = 54935
Input 5 - 1590695523 * -146190406 = 9146
Input 6 - -498698720 * -2067076911 = -36140
Input 7 - 313980746 * -1855353564 = -8629
Input 8 - 927455040 * 909339715 = 34099
Input 9 - 1879963342 * 1046709913 = -42999
Input 10 - -952410623 * 1623673218 = 22477
Input 11 - 1749807198 * 1869257233 = -3388
Input 12 - -990812923 * -1437499032 = -9642
Input 13 - -1743332514 * -851890251 = -3255
Input 14 - -229032939 * 145030982 = -4842
Input 15 - 1832567514 * -2056752860 = 22395
Input 16 - 282093348 * -1241046311 = 2370
Input 17 - 1565880994 * -484387807 = 353
Input 18 - 1999420519 * -1349377428 = 3037
Input 19 - 1267338533 * 1385074179 = -10953
Input 20 - 1497917785 * 1733168506 = 27006
Input 21 - 1953134459 * -635395530 = -5560
Input 22 - -99251095 * -1201396468 = -9460
Input 23 - -1789185521 * -60844315 = 2344
Input 24 - -652639821 * -116628051 = -3047
Input 25 - -1658301030 * 2025820983 = 15445
Input 26 - -1200444634 * 1263628989 = 39982
Input 27 - 1761762889 * 619516577 = -1501
Input 28 - -469035308 * -1967811798 = -21223
Input 29 - 1055432037 * -1159144932 = -13046
Input 30 - 203104764 * 1388044452 = -25119
Input 31 - -2050287947 * -287657486 = -14380
Input 32 - 1364308297 * 1588044244 = -19622
Input 33 - -668444521 * 1385425608 = 2960
Input 34 - -524577492 * 561837058 = 28880
Input 35 - -2068650633 * 128636962 = -19498
Input 36 - 483011732 * 1182803727 = -19960
Input 37 - -458839202 * 1950117664 = 58397
Input 38 - -1078623433 * -623901530 = -20372
Input 39 - 32477355 * 1786470108 = 36883
Input 40 - -924845581 * 509255833 = -22863
Input 41 - -1651858733 * 1231316961 = 35406
Input 42 - 18767643 * 299903318 = 5274
Input 43 - -447955336 * -264545318 = 13550
Input 44 - -524096024 * -359393064 = -29472
Input 45 - 1172687165 * -673338130 = -28826
Input 46 - -127350075 * -1774661321 = -19123
Input 47 - 493451948 * -719182290 = 14871
Input 48 - 732893559 * 1684471745 = 14636
Input 49 - 1572741726 * 1959349349 = 12416
Input 50 - 1736234271 * 1725470179 = -18144
Input 51 - -723082767 * 1112350300 = 58704
Input 52 - -770545660 * -1922770595 = -26034
Input 53 - 1944778137 * -1863275624 = -53361
Input 54 - 1556815405 * 345083032 = -28853
Input 55 - 1925585144 * 1391797504 = 8058
Input 56 - -2107745853 * -363544423 = -20745
Input 57 - 151264043 * 1232408018 = -502
Input 58 - -1394276302 * -1040921752 = -8808
Input 59 - 1181093746 * 1571484118 = -7275
Input 60 - 1192575604 * 74971289 = -9693
Input 61 - 2146758244 * -1180877586 = -20693
Input 62 - 2079677389 * -444125406 = -21719
Input 63 - -1687006335 * 1878731255 = 18877
Input 64 - 566795218 * 76909930 = -16833
Input 65 - -1420532271 * 1950482513 = 42309
Input 66 - 1707958342 * -757420128 = -26932
Input 67 - 500023378 * -1720646393 = 21712
Input 68 - 1344051070 * 2140694887 = 36081
Input 69 - 2147215850 * -1404302226 = 36944
Input 70 - 1948951637 * -373446303 = -11505
Input 71 - 464899997 * -1143247050 = 22895
Input 72 - -97054943 * 2132691636 = 32002
Input 73 - -2068843520 * -1879556868 = -1676
Input 74 - -865608691 * -1645149433 = -20990
Input 75 - 737796484 * 359495035 = 70605
Input 76 - -930037506 * -705487466 = -31828
Input 77 - 743520928 * 1839675886 = -1987
Input 78 - 1477664966 * 1631862215 = 4398
Input 79 - 725439180 * -1566557596 = 15258
Input 80 - -1434137957 * 402544254 = 40548
Input 81 - 1608613759 * 1856270065 = 1815
Input 82 - -1634828478 * -702731626 = -4180
Input 83 - -134138619 * 125810070 = -2642
Input 84 - 1866908966 * -1960354557 = 8178
Input 85 - -1117448149 * -674545805 = -6645
Input 86 - 518422069 * 1690481352 = -24056
Input 87 - 1495883903 * -1605354437 = 7989
Input 88 - -1806481308 * 106355124 = -2382
Input 89 - 2052708446 * 1249537117 = 51956
Input 90 - 1956547278 * -297617452 = -10798
Input 91 - 796460037 * 794212634 = 12461
Input 92 - 749428652 * -2101694951 = -21004
Input 93 - -2032996521 * -2005037681 = -27146
Input 94 - 2019112651 * 1129473472 = 7800
Input 95 - 1225964212 * 294918828 = -10987
Input 96 - 852027437 * -901230162 = 26802
Input 97 - 1995465699 * -232411077 = 12701
Input 98 - 903918819 * -1806061130 = -16806
Input 99 - -2077446604 * -4430843 = 14536
Input 100 - -2138267674 * -328125550 = -3931
Input 101 - 1942487345 * -350988207 = 5642
Input 102 - -2118340710 * -761722129 = -26166
Input 103 - 1106146040 * -18203545 = 27018
Input 104 - 1607134236 * -33593563 = 24857
Input 105 - -1075072324 * 1111368634 = 20246
Input 106 - 1672327211 * 1698492678 = 17997
Input 107 - 820533319 * -105888090 = -19126
Input 108 - -965081840 * 635856997 = 6290
Input 109 - -1008560635 * -243058909 = -22750
Input 110 - -1814138460 * -1836805866 = -52633
Input 111 - -132681162 * 1017723788 = 13341
Input 112 - 229534031 * 726555332 = 18841
Input 113 - -616135652 * -431778256 = -13652
Input 114 - 80004865 * 519386061 = 8062
Input 115 - 1663082212 * -57262902 = -6620
Input 116 - -1141510189 * 634911918 = -21384
Input 117 - 276590231 * 953250452 = -8700
Input 118 - 1779349259 * 422401842 = 19751
Input 119 - 828021828 * -809160212 = -1427
Input 120 - 672462341 * -88733881 = 357
Input 121 - -1809360036 * 1016502499 = -7386
Input 122 - -729073012 * -139754223 = -18477
Input 123 - 1029801201 * -459379827 = -24805
Input 124 - 2109228175 * -970291861 = 7415
Input 125 - -777035140 * 1486890148 = -6288
Input 126 - 768696856 * -45452091 = 25283
Input 127 - -1927818396 * 391167683 = -14943
Input 128 - -1100234120 * -673847637 = -27635
Input 129 - -215642073 * 1736192460 = 34645
Input 130 - 1268168767 * 1950639645 = 39495
Input 131 - -405091965 * 165633897 = 27735
Input 132 - -291764096 * 1264897266 = 31242
Input 133 - -2044151386 * 1440568754 = 12767
Input 134 - 604760474 * 1827298563 = 25777
Input 135 - -1886297933 * 1630632026 = 48606
Input 136 - 1867240401 * -111505996 = -34176
Input 137 - -1455636642 * -710818907 = -33172
Input 138 - -1821877803 * 1560255364 = -21981
Input 139 - -895941706 * 1818309168 = 45543
Input 140 - 1957252322 * 1306709618 = 25826
Input 141 - -1712570395 * -673553369 = -10
Input 142 - 1958537789 * 1321601416 = -7070
Input 143 - 1564553869 * 1199988245 = 6334
Input 144 - 1525441353 * -335333091 = 6668
Input 145 - -668549803 * 347862435 = -8286
Input 146 - -763632272 * 179635108 = -16742
Input 147 - -927314922 * -275260362 = -24478
Input 148 - -437760741 * 1374955928 = 16114
Input 149 - 2120166227 * 1060456261 = 21753
Input 150 - 1287909581 * -2118847621 = -2377
Input 151 - 209952823 * -1083734163 = -652
Input 152 - -2107214019 * 229560472 = -18730
Input 153 - 983433585 * 796338685 = 21230
Input 154 - -1407069121 * -2139013190 = -17261
Input 155 - -1357777164 * -978714807 = 5762
Input 156 - -878619358 * 2060473740 = 20319
Input 157 - 1276945709 * -261621699 = 3093
Input 158 - 1033009048 * 1522912483 = -13720
Input 159 - 1672225559 * -2071809851 = -38105
Input 160 - -587696016 * 90545903 = 12484
Input 161 - -101035032 * 585730300 = -8172
Input 162 - -1732548853 * 420178156 = 16349
Input 163 - 1099909766 * -566989443 = 30042
Input 164 - -1670652689 * -2047431671 = -36587
Input 165 - -912777307 * -417264827 = 11717
Input 166 - 686535488 * -1304503860 = 1141
Input 167 - -1091819657 * -641433223 = -12124
Input 168 - 1858127193 * 1851569016 = 25275
Input 169 - 945511383 * -1912690252 = -30879
Input 170 - -975930738 * -676909780 = -15207
Input 171 - 158167437 * -137653850 = -31091
Input 172 - 1975563657 * -1467782253 = -29862
Input 173 - -1655654211 * 2043689522 = 35279
Input 174 - 350712935 * -355129419 = -13574
Input 175 - 396422523 * -934924216 = 30144
Input 176 - -545540557 * 363643022 = -23111
Input 177 - -295069241 * 1791323420 = 41282
Input 178 - -2128389150 * -1758564176 = -19823
Input 179 - -263569057 * -1538852048 = -12665
Input 180 - 1955383838 * -641957538 = -36878
Input 181 - -1979587308 * 1395793571 = 16514
Input 182 - 1498100227 * 1414799707 = 15699
Input 183 - 286835136 * 1327061486 = 15664
Input 184 - -277077251 * 1087006891 = -9369
Input 185 - -205237040 * -879770263 = -18143
Input 186 - 1673518703 * -1825793551 = -26918
Input 187 - 1776379219 * 1251842231 = -31068
Input 188 - -1294719328 * -185585844 = 8418
Input 189 - 238403481 * -745514998 = 11772
Input 190 - 1416070045 * -375834187 = -16849
Input 191 - -269143771 * 715325166 = -13756
Input 192 - -842593170 * 1280339638 = 24680
Input 193 - 203951749 * 989395231 = 4401
Input 194 - 1549180473 * 96008691 = -7579
Input 195 - 1802324022 * -116750763 = -3853
Input 196 - 85707792 * -608723437 = -21844
Input 197 - -2102837299 * -425688530 = -18112
Input 198 - -1535901964 * 953355575 = 20731
Input 199 - 639138082 * -1434107167 = 712
PASSED